Search revamp

sagotch opened the pull request:
Les indexes doivent être régénérés pour que la nouvelle recherche fonctionne:
```
gwfixbase.exe -index /path/to/base.gwb
```

Cette refonte apporte les modifications suivantes à la recherche:
- par défaut, la recherche cherche en découpant les noms/prenoms sur les mots:
  e.g. Julien Romain Edouard SAGOT va correspondre à une recherche sur "Julien", sur "Edouard", et sur "Romain".
- La recherche avancée permet de faire une rechercher "exact": la decomposition du nom/prenom doit correspondre exactement a la decomposition recherche (pas un mot de plus, pas une mot de moins)
- La recherche sur "Julien Romain" fera ressortir "Julien Romain Edouard" pour la recherche avancée (sauf si recherche exacte est activée). Par contre en recherche simple, cela n'est pas encore effectif, et "Julien Romain" ne verra pas ressortir "Julien Romain Edouard" (ne ressort de toute façon pas dans la recherche actuelle)
- les nom/prenom aliases sont pris en compte au même titre que les nom/prenom

---

Indexes need to be regenerated in order to make this work:
```
gwfixbase.exe -index /path/to/base.gwb
```

Copy/paste from https://github.com/geneweb/geneweb/issues/904

> I will only talk about first names and surnames, but alias <del>and nicknames</del> should probably be treated the same way
>
>     When searching, first names are splitted on space
>     When searching, surnames are splitted on space and hyphen
>     Search is accent insensitive (ascii versions of the string is used, see source of "unidecode" library)
>     Search is case insensitive
>     Order of names does not matter
>     If a person match each component of the search, it is selected (if first name is "julien romain edouard" and if we search "julien edouard", it will be selected)
>     When performing "exact match" (new option) search, searching "julien romain edouard" will not match "julien romain". "julien romain" will match "julien romain" and "romain julien" but not "julien romain edouard"
> 

---

Please answer this thread if you spot a bug

---

Simple search for first name "rené"

![image](https://user-images.githubusercontent.com/1826552/84671540-cb713200-af27-11ea-8143-0b4918721d45.png)
Note: values matching exactly the search will be placed as first item, the rest is sorted alphabetically.

---

When clicking on a precise value:
![image](https://user-images.githubusercontent.com/1826552/84671710-fd829400-af27-11ea-9a4b-0cd004305a26.png)

---

Advanced search for "Julien Romain":
![image](https://user-images.githubusercontent.com/1826552/84671847-21de7080-af28-11ea-82c4-7dc441a3eba4.png)

---

Avanced search for "exactly Julien Romain":
![image](https://user-images.githubusercontent.com/1826552/84671936-3d497b80-af28-11ea-8596-a5c98b96607f.png)

---

TODO (non bloquant pour merge): Par contre en recherche simple, cela n'est pas encore effectif, et "Julien Romain" ne verra pas ressortir "Julien Romain Edouard" (ne ressort de toute façon pas dans la recherche actuelle)

---

À noter : les pages de recherche simple ainsi que la page des nom et prénoms se retrouvent aussi affectées (enrichit par les aliases). À voir si cela ne charge pas trop les listes et s'il ne faut pas améliorer cette partie

Commit:
gwdb-legacy: fixed persons_of_{sur,first_}name in order to include aliases

Commit:
Optimized gen_person_misc_names

Commit:
Added Name.concat function

Commit:
gwdb-legacy: optimize make_name_index

Commit:
Added Name.split_{f,s}name functions

Commit:
Replaced Mutil.surnames_pieces with Name.split_sn

Commit:
gwdb-legacy: factorized index generation

Commit:
gwdb-legacy: optimize make_strings_of_fsname

Commit:
gwdb-legacy: faster make_strings_of_fsname

Commit:
Added Mutil.split_{f,s}name_callback functions

Commit:
gwdb-legacy: used Name.split_{f,s}name_callback

Commit:
Drop support for GeneWeb < 4.10 version indexes

Commit:
gwdb-legacy: added Dutil.name_index

Commit:
gwdb-legacy: strings of fsname: do not filter particles

Commit:
Removed Gutil.arg_list_of_string

Commit:
Removed Gutil.trim_trailing_spaces

Commit:
Removed unused code

Commit:
Removed all the alphabetic_* stuff, replaced by Utf8.compare

Commit:
Removed Gutil.{father,mother,parent_array} functions

Commit:
gwdb-legacy: fixed strings_of_fsname

Commit:
Enhanced advSearchOk

Commit:
Added 'exact' checkbox to advanced search

Commit:
gwdb-legacy: fsname index: add aliases

Commit:
Gwdb: match_{sur,first_}name now search in aliases if asked for

Commit:
AdvSearch: search in first names/surnames aliases.

Commit:
gwdb-legacy: add proper support for aliases when building indexes.

Commit:
m=N&tri=S: fixed surname sort

Commit:
gwdb-legacy: fixed surnames_aliases in indexes

Commit:
gwdb-legacy: include unsplitted value to strings_of_fsname index.

Commit:
Fixed url for m=N (t=A instead of t=N)

Commit:
Moved and improved 'ancestor line grouping' from Some to Util.

Commit:
Some and SearchName modules revamped.

Commit:
Fixed 063f29599 by assuring retro-compatibility of strings index

Commit:
gwdb-legacy/strings_of_fsname: fixed missing aliases handling

Commit:
gwdb-legacy: fixed 'find' in persons_of_first_name_or_surname

Commit:
Update lexicon

Commit:
Fixed 'unused value' warnings

Commit:
Some.persons_of_fsname -> Some.new_persons_of_fsname

Commit:
Compatibility for OCaml < 4.08.0

Commit:
Some.persons_of_absolute_aux: also support aliases

hgouraud commented:
```
iMac-H:geneweb Henri$ ocaml configure.ml 
iMac-H:geneweb Henri$ make clean distrib
-n Cleaning...
 Done!
-n Generating hd/etc/version.txt...
 Done!
-n Generating lib/gwlib.ml...
 Done!
-n Generating bin/setup/setup.ml...
 Done!
-n Generating lib/templ.ml...
 Done!
-n Generating benchmark/dune...
 Done!
-n Generating bin/dune...
 Done!
-n Generating lib/dune...
 Done!
-n Generating test/dune...
 Done!
cat dune-workspace.in | sed  -e "s/%%%DUNE_PROFILE%%%/dev/g" > dune-workspace
dune build 
File "lib/gwdb-legacy/outbase.ml", line 50, characters 26-29:
Error: Unbound module Int
make: *** [exe] Error 1  
iMac-H:geneweb Henri$ 

```

dominique95880 commented:
Sous **Windows,** ça compile bien
adm0018@Derdo01 /home/geneweb/geneweb
$ git checkout search-revamp
La branche 'search-revamp' est paramétrée pour suivre la branche distante 'search-revamp' depuis 'origin'.
Basculement sur la nouvelle branche 'search-revamp'

adm0018@Derdo01 /home/geneweb/geneweb
$ ocaml ./configure.ml

adm0018@Derdo01 /home/geneweb/geneweb
$ make clean distrib
Cleaning... Done!
Generating hd/etc/version.txt... Done!
Generating lib/gwlib.ml... Done!
Generating bin/setup/setup.ml... Done!
Generating lib/templ.ml... Done!
Generating benchmark/dune... Done!
Generating bin/dune... Done!
Generating lib/dune... Done!
Generating test/dune... Done!
cat dune-workspace.in | sed  -e "s/%%%DUNE_PROFILE%%%/dev/g" > dune-workspace
dune build
rm -f -r distribution
mkdir distribution
...
cp -R hd/* distribution/gw/
rm -f distribution/exe-version.txt
echo "Commit: `git log  -1 | grep commit | cut -c8-15`, `date`"      > distribution/commit.txt
echo "`ocaml  -version`"      >> distribution/commit.txt
echo "`camlp5 -v 2>&1`"       >> distribution/commit.txt
echo "-----"                  >> distribution/commit.txt

Mais, à l'exécution, si je demande les prénoms Martin "tout court", (case Exact non cochée) je n'ai que les Martin "tout court" (idem si Exact est cochée)
Requête évoluée

Recherche tous les Martin,.

    Martin COUDEVILLE ca 1590-1635..1641, fils de 1er, marié avec Antoinette SERGEANT en 1615.
    Martin MORMENTYN 1633-1688, fils de Pierre et Martine PALYNCK, marié avec Catherine COUDEVILLE en 1667.
    Martin VANHERSECKE ca 1806-1859/.
    Martin DUBOIS ca 1801-1853/.
    Martin WIPPELIER ca 1775-.
    Martin YVOZ ca 1806-1867/, marié avec Marie Octavie JAY en 1842.
etc.

tandis que si je demande les prénoms Martin" tout court" et NOM COUDEVILLE, j'ai bien les Martin "n'importe où" COUDEVILLE.
Requête évoluée

Recherche tous les Martin COUDEVILLE,.

    Martin COUDEVILLE ca 1590-1635..1641, fils de 1er, marié avec Antoinette SERGEANT en 1615.
    Pierre Martin COUDEVILLE 1786-1824, fils de Mathieu et Jacqueline PIERENS, marié avec Sophie Isabelle DURANEL en 1817.
    Louis Henri Martin COUDEVILLE 1845-1845, fils de Charles Louis Benoît et Émélie Sophie Reine RENIER.
    Louis Jacques Martin COUDEVILLE 1838-1840, fils de Barthélémy Séverin Cornil et Jeanne Cécile DENYS.
    Pierre Martin COUDEVILLE 1790-1813, fils de Maurice et Marie Barbe Josèphe PORREYE.
    Benoît Martin Louis COUDEVILLE 1827-, fils d’Augustin Louis et Angélique Victoire VERIEPE.
    Martin Jean COUDEVYLLE 1798-1799, fils de Benoît Constantin et Marie Catherine DEBEYER.

Dominique

sagotch commented:
Pour le probleme de compilation c'est un probleme de OCaml < je sais plus quelle version. Je vais corriger.

Pour les martins les indexes ont bien été regenerés avant test ? Si oui alors il y a un bug effectivement, je regarde ca demain matin.

Merci pour les retours :+1:

dominique95880 commented:
"Pour les martins les indexes ont bien été regenerés avant test ? Si oui alors il y a un bug effectivement, je regarde ca demain matin."

Oops ! non, mea culpa.
Je vais faire un nouveau test dans la journée.

Dominique

Edit : après avoir créé une base à partir d'un fichier gw, c'est bon pour les Martin "n'importe où"
Requête évoluée

Recherche tous les Martin,.

    François Barthélémy Martin CALLIEU 1816-1890, fils de Laurent Winoc et Jeanne Thérèse GHYS.
    Cornil Ferdinand Martin DERAM 1816-1879, fils de Ferdinand Cornil et Anne Thérèse Jacqueline LAFORCE, marié avec Martine Sophie DEBREYNE en 1847.
    Pierre Martin MORTIER 1724-1808, fils d’André et Marie Jeanne BECUWE, marié avec Marie Pétronille DONAES en 1750.
    Pierre Martin DERAM 1805-1855, fils de Ferdinand Cornil et Anne Thérèse Jacqueline LAFORCE, marié avec Winoc Sophie BERTEIN en 1827.
    Benoît Martin Alphonse BOGAERT ca 1847-, marié avec Lucie Julia DECLERCK en 1886.
    Gustave Martin Cornil DEFRAEYE ca 1889-, fils de Louis François Gustave et Léonie Eugénie D'HOËRE.
    Martin HORLAIT 1985, fils de Stéphane et Frédérique DURIEZ.
    Martin FRANCKE ca 2000, fils de Stéphane et Véronique COUDEVILLE.
    Martin COUDEVILLE ca 1590-1635..1641, fils de 1er, marié avec Antoinette SERGEANT en 1615.
...

Commit:
Fixed compilation OCaml < 4.08

Commit:
Fixed compare_place test.

[compare] returned value may be other than 0, -1, 1:
utop # Char.compare 'a' 'o' ;;
- : int = -14

Commit:
gwdb-legacy: fixed make_strings_of_fsname

Commit:
Fixed simple search

hgouraud commented:
Pour la recherche sur les noms et prénoms, il serait opportun d'utiliser l'affichage proposé par `print_alpha_list`.
Dans le cas de la recherche par prénom, il serait préférable d'afficher la liste sous la forme Surname, First_name. Les fonctions `person_text` pourraient avoir un paramètre `fn_sn` précisant l'ordre d'affichage.
Dans la mesure ou les prénoms alias font partie de la recherche, comment pourrait on afficher l'alias plutôt que le prénom si c'est lui qui a provoqué le hit. Le résultat du search pourrait être une liste de `(personnes * reason )`  avec reason précisant `Fname, FnameAlias, Surname, SurnameAlias, Alias, PublicName, Sobriquet, ...`; La fonction d'affichage pourrait alors choisir le bon terme.

sagotch commented:
Tout à fait d'accord, l'affichage mérite d'être retravaillé, reste à savoir comment. En attendant, tant que c'est fonctionnel ça sera bon pour moi comme première étape.

J'ai deux templates dans mes cartons pour les recherches simples, pas encore tout à fait prêts. Mais je les pousserai bientot pour revue et discussion.

sagotch commented:
Effets de bords indésirables, de partout, problème de perfs sur les gros arbres -> poubelle
